trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedGhostUpdate.java

```diff
@@ -0,0 +1,221 @@
+		final ExpressionResultBuilder builder = new ExpressionResultBuilder();
+		for (final String st : mStatements) {
+			ACSLNode acslNode = null;
+			try {
+				acslNode = Parser.parseComment("lstart\n ghost " + st, getStartline(), 1);
+			} catch (final ACSLSyntaxErrorException e) {
+				throw new UnsupportedSyntaxException(loc, e.getMessageText());
+			} catch (final Exception e) {
+				throw new AssertionError(e);
+			}
+			builder.addAllExceptLrValue((ExpressionResult) dispatcher.dispatch(acslNode, mMatchedAstNode));
+		}
+		return builder.build();
+	}
+
+	private String getNameOfCalledFunction() {
+		final IASTExpression expr;
+		if (mMatchedAstNode instanceof IASTExpression) {
+			expr = (IASTExpression) mMatchedAstNode;
+		} else if (mMatchedAstNode instanceof IASTExpressionStatement) {
+			expr = ((IASTExpressionStatement) mMatchedAstNode).getExpression();
+		} else {
+			return null;
+		}
+		if (expr instanceof IASTFunctionCallExpression) {
+			final IASTExpression function = ((IASTFunctionCallExpression) expr).getFunctionNameExpression();
+			if (function instanceof IASTIdExpression) {
+				return ((IASTIdExpression) function).getName().toString();
+			}
+		}
+		return null;
+	}
+
+	private static List<Statement> annotateAtomicCall(final ILocation loc, final List<Statement> programStatements,
+			final List<Statement> ghostUpdate, final String functionName) {
+		final List<Statement> result = new ArrayList<>();
+		boolean isAnnotated = false;
+		for (final Statement st : programStatements) {
+			if (isAnnotated || !(st instanceof AtomicStatement)) {
+				result.add(st);
+				continue;
+			}
+			final Statement[] atomicBody = ((AtomicStatement) st).getBody();
+			if (Arrays.stream(atomicBody).anyMatch(
+					x -> x instanceof CallStatement && ((CallStatement) x).getMethodName().equals(functionName))) {
+				isAnnotated = true;
+				result.add(new AtomicStatement(loc,
+						DataStructureUtils.concat(ghostUpdate.toArray(Statement[]::new), atomicBody)));
+			} else {
+				result.add(st);
+			}
+		}
+		if (!isAnnotated) {
+			throw new UnsupportedOperationException("No statement found to annotate with the expected ghost update");
+		}
+		return result;
+	}
+
+	private static List<Statement> annotateFork(final ILocation loc, final List<Statement> programStatements,
+			final List<Statement> ghostUpdate) {
+		final List<Statement> result = new ArrayList<>();
+		boolean isAnnotated = false;
+		for (final Statement st : programStatements) {
+			if (!isAnnotated && st instanceof ForkStatement) {
+				isAnnotated = true;
+				result.add(new AtomicStatement(loc, ghostUpdate.toArray(Statement[]::new)));
+			}
+			result.add(st);
+		}
+		if (!isAnnotated) {
+			throw new UnsupportedOperationException("No statement found to annotate with the expected ghost update");
+		}
+		return result;
+	}
+
+	@Override
+	public ExpressionResult transform(final ILocation loc, final IDispatcher dispatcher,
+			final ExpressionResult expressionResult) {
```

Comment:
This method defines the points where we allow ghost updates. We may want to extend or completely change these points in the future (e.g., allowing updates at sequence points). Should we do anything to ease such changes later (maybe add a setting, or at least a boolean flag, ...) ?

Reply:
You are right, the behavior of ghost updates might can. But what should happen, if the setting/flag is set to true/false?

Reply:
Hm, I can't think of something meaningful here either. Disregard my comment.